\begin{abstract}
    There is a rapid growth in the number of available biological datasets due
    to the advent of high-throughput data collection instruments combined with
    cheap compute infrastructure.  Datasets from modern instruments
    enable analysis of biological data at different levels, from small DNA
    sequences through larger cell structures, and up to the function of entire
    organs. 
    These new datasets have brought the need for the development of new software
    tools and packages to enable novel insights to the underlying biological
    mechanisms, in the development, and progression of, diseases such as cancer. 
    
    The heterogeneity of biological datasets require researchers to tailor
    the exploration and analyses using a range of different tools and systems.
    However, despite the need for their integration, few of them provide
    standard interfaces for analyses implemented using different programming
    languages and frameworks. In addition, because of the many tools, input
    parameters, and reference to databases, it is necessary to record these
    correctly. The lack of such details complicates the reproducing of original
    results and reuse of the analyses on new datasets. This increases time
    for analysis and leaves unrealized potential for scientific insights.
 
    This dissertation argues that we can develop unified systems for
    reproducible exploration and analysis of high-throughput biological
    datasets.  We propose an approach, \Glsentryfirstplural{sme}, for developing
    data analysis pipelines and data exploration applications in cancer
    research.  We realize \glspl{sme} using software container technologies
    together with well-defined interfaces, configuration, and orchestration. It
    simplifies the development of such applications, and provides detailed
    information needed to reproduce the analyses.
    
    Through this approach we have developed different applications for analyzing
    high-throughput \gls{dna} sequencing datasets, and exploring gene expression
    data integrated with questionnaires, registry, and online databases. The
    evaluation shows how we effectively capture provenance in analysis pipelines
    and exploration applications. This simplifies reproducibility through the
    sharing of methods, tools, and data, and is fundamental to science. 
    
\end{abstract}

