\begin{abstract}
    There is a rapid growth in the number of available biological datasets due
    to the advent of high-throughput data collection instruments together with
    cheap computational infrastructures.  Datasets from modern instruments
    enable analysis of biological data at different levels, from small DNA
    sequences through larger cell structures, and up to the function of entire
    organs.  The heterogeneity of these different biological datasets has led to
    the development of a wealth of software packages and systems for researchers
    to use in data exploration and analysis.  This leads to the potential for
    novel insights to the underlying biological mechanisms in the development
    and progression of diseases such as cancer. 
    
    To discover and study patterns in biological datasets researchers tailor the
    exploration and analyses using a range of different tools and systems.
    However, despite the need to use a range of tools, few of these provide
    standard interfaces for analyses implemented in different programming
    languages and frameworks. In addition, because of the long list of tools,
    input parameters, and reference to databases it is difficult to correctly
    report the correct details of an analysis. The lack of such details can
    complicate the reuse of analyses for other datasets, or even reproducing the
    results of the original data. This increases both analysis time and leaves
    unrealized potential for scientific insights.

    This dissertation argues that, instead, we can develop unified systems for
    reproducible exploration and analysis high-throughput biological datasets.
    We propose an approach, Small Composable Units (SCUs), that orchestrates the
    execution of analysis pipelines and data exploration applications. We
    realize this approach using software container technologies together with
    well-defined interfaces, configuration, and orchestration.  It simplifies
    the development of such applications, and provides correct and detailed
    information to reproduce the analyses.
    
    We demonstrate different applications for analyzing high-throughput
    \gls{dna} sequencing datasets, and exploring gene expression data integrated
    with questionnaire, registry, and online databases. Our evaluation shows how
    we effectively capture provenance in analysis pipelines and exploration
    applications to simplify reproducibility and encourage sharing of methods
    and tools.
    
\end{abstract}

