\begin{abstract}
    There is a rapid growth in the number of available biological datasets due
    to the advent of high-throughput data collection instruments together with
    cheap computational infrastructures.  Datasets from modern instruments
    enable analysis of biological data at different levels, from small DNA
    sequences through larger cell structures, and up to the function of entire
    organs.  The heterogeneity of these different biological datasets has led to
    the development of a wealth of software packages and systems for researchers
    to use in data exploration and analysis.  This leads to the potential for
    novel insights to the underlying biological mechanisms in the development
    and progression of diseases such as cancer. 
    
    Researchers tailor the exploration and analysis of their datasets using a
    range of different tools and systems. However, because of the specialized
    nature of each step of the analysis, few of the tools provide useful
    interfaces for analyses across programming languages and frameworks. In
    addition, reporting the details of an analysis becomes a tedious task
    because of the long list of different tools, input parameters and reference
    to databases. The lack of such details can complicate reusing the analyses
    for other datasets, or even reproducing the results of the original data.
    This increases both analysis time and leaves unrealized potential for
    scientific insights.

    This dissertation argues that, instead, we can design systems for
    exploring and analyzing high-throughput biological datasets from small
    composable pieces. In particular we show the viability through software
    container technologies together with well-defined interfaces, configuration,
    and orchestration. This approach simplifies the development of analysis
    pipelines and interactive data exploration applications. In addition it
    provides the necessary information to reproduce analyses and share the
    methods across teams. 
    
    We show the feasibility of our approach through applications for analyzing
    high-throughput \gls{dna} sequencing datasets, and exploring gene expression
    data integrated with questionnaire, registry, and online databases. Our
    evaluation shows how we effectively capture provenance in analysis pipelines
    and exploration applications to simplify reproducibility and encourage
    sharing of methods and tools.
    
\end{abstract}

