% observation
% LAB: from the -> due to the?
There is a rapid growth in the number of available biological datasets from the
decreaseing cost of data collection. 
% LAB: the opportunity -> opportunities (mange små dataset)
This brings the opportunity to gain novel
% LAB: blir "in the" rett?
insights in the underlying biological mechanisms in the development and
progression of diseases such as cancer. 
& LAB: The gorwing number -> The wide range / variability / noe-annet of all the 
The growing number of datasets has led
% LAB: availability -> development
to the availablility of a wealth of software packages and systems to explore and
analyze these datasets. 
% LAB: systems -> tools
However, these systems typically target only small
portions of the deep analysis pipeline required to transform the raw data into
interpretable results. While the tools are used to provide novel insights in
diseases, there is little emphasis on reporting and sharing information about
tool versions, input parameters, and other information that can help others use
the same methods on their own datasets. This leads to unneccesary difficilties
% LAB: to ganger This leads
% LAB: potential -> potential for scientific insights and commerciial
% utlization...
to reuse known methods, and difficulties in reproducing analyses, which leads
to unrealized potential in the growing number of available datasets. 

% challenge

There are several challenges in enabling 
% LAB: efficient blir uklart her. Det kan være både relatert til penger, CPU
% cycles, og manuelt arbeid. Og det er en viktig setning siden vi her setter
% premissen for thesis.
efficient analysis and exploration of
biological datasets. 
The first is deep analysis pipelines: it is too time consuming to set up 
% LAB: an -> a new; to start analyzinng -> for a new analysis
an
analysis pipeline to start analyzing a biological dataset. 
% LAB: dette ble sagt i forrige paragraf
Little effort is put
into sharing full pipelines from raw data to interpretable results.
% LAB: ensuring for hva? For subsequent analyses?
% LAB: challenge er kanskje ikke "correct tool...", men det som skal ensures som
% krever at det er "correct tool..."
The second challenge is ensuring correct tool, tool versions, databases, and
datasets. Deep analysis pipelines use multiple tools from different developers
to transform the data into interpretable results. Modifying any of these will
% LAB: final -> pipeline output
impact the final result and must be recorded to enable reproducible analyses. 
% LAB: ikke sikkert at leseren skjlnner hva som er challenge. 
A third challenge is the large variation in study design, datatypes,
% LAB: in question -> under study
and biological functions in question.
% LAB: ikke sikkert at leseren skjønner challenge her
The final challenge is interactive applications: end user-facing application
for exploring results of statistical analyses are often decoupled from the
analyses, simply visualizing a static dataset.  

% from the above list
As a result, there are a wealth of approaches and systems to enable analysis of
the complex biological data. To develop analysis pipelines, Galaxy\cite{galaxy}
has for a long time provided a simple interface to set up and execute analysis
pipelines for genomic datasets. However, the Galaxy system is less effective for
explorative and flexible analyses.\cite{spjuth2015experiences}
Pachyderm\cite{pachyderm} is a system for developing more flexible analyses that
support comparison of pipeline runs from different workflow configurations and
datasets. However it has yet to see wide-spread adpotion in Bioinformatics.
New initiatives such as the \gls{cwl} provide users a standardized way of
describing an analysis pipeline and has multiple implementations such as
the reference implementation
cwl\_runner,\footnote{\url{github.com/common-workflow-language/cwltool}}
Arvados,\cite{arvados} Rabix,\cite{rabix} Toil,\cite{toil} Galaxy,\cite{galaxy}
and AWE.\cite{awe} 


With new analysis systems and frameworks continuing to be implemented every
year, it seems that enabling reproducible analyses of high-throughput datasets
is still an area that require thorough consideration and developer effort.

\begin{enumerate*}[label=(\roman*)]
    \item cwl,cwl\_runner, Arvados, Rabix, Toil, Galaxy,
        AWE;   
    \item pachyderm
    \item wrt. study design, datatypes, 
    \item rstudio, shiny, opencpu,
\end{enumerate*} 


\emph{Our solution}: 
This dissertation argues that, instead, ... 
% unified/hollistic: unified approach/model 
% kvik/walr (data straming/ of open source systems including Apache Spark as the
% common foundation; Shark for SQL processing; and Spark Streaming for
% distributed stream processing) 
% nowac-pakken 

% bruk challenges for å si hvorfor vår er bedre
Our analysis solution has the benefit that it integrates with modern version
control systems to provide provenance information on datasets. It also runs on a
wide range of software platforms and is targeted to the compute infrastructure
found in hospital environment. 

Our solution to build data explorations provides the benefit of developers being
able to develop applications in any programming language. It also facilitates
the reuse and interface with the wealth of statistical packages directly from a
user-interface rather than analysis scripts. 

The resulting approach as several key advantages over current systems: 
\begin{itemize} 
    \item
\end{itemize} 

We implement walrus/kvik... 

% We implement the RDD architecture in a stack of open source systems including
% Apache Spark as the common foundation; evaluation: 
From a longer-term perspective, ....

\emph{Thesis statement}:
A hollistic approach based on ...containers ... efficient analsysis of diverse
biological datasets ... 
% hva hollistic can ... support diverse distr. comp.
% A holistic approach to data analysis and exploration of high-throughput
% dataasets in bioinformatics??? 

\section{Problems with Data Analysis and Exploration in Bioinformatics} 
    List of things that we want to fix. 

%  no way I'm keeping the acronym    
\section{The Container-based Data Analysis Model (CDAM)} 
    The Solution (?) 

\section{Use of the CDAM} 

To provide an overview of the statistical analyses from raw data to
interpretable results we propose our tool \emph{walrus}. It lets users create
and run analysis pipelines in bioinformatics to e.g. analyze high-throughput
sequencing datasets. In addition, it tracks full provenance of the input,
intermediate, and output data, as well as tool parameters. With \emph{walrus} we
have successfully built analysis pipelines to detect somatic mutations in breast
cancer patients. 

To develop applications that interface with the underlying statistical analyses
we have built \emph{Kvik}. Kvik allows applications written in any modern
programming language to interface with the wealth of bioinformatics packages in
the R programming language, as well as information available through online
databases. We have used Kvik to develop the MIxT system for exploring and
comparing transcriptional profiles from blood and tumor samples in breast cancer
patients. 


\textbf{walrus}
\textbf{kvik} 
\textbf{mixt} 

\section{Summary of Results} 

\section{List of papers} 
\begin{itemize}
    \item
        \emph{Kvik: three-tier data exploration tools for flexible analysis of
        genomic data in epidemiological studies}
        \\
        \textbf{Bjørn Fjukstad}, Karina Standahl Olsen, Mie Jareid, Eiliv Lund,
        Lars Ailo Bongo. 
        \\ 
        F1000Research 2015.
        
    \item 
        \emph{Building Applications For Interactive Data Exploration In Systems
        Biology.}
        \\
        \textbf{Bjørn Fjukstad}, Vanessa Dumeaux, Karina Standahl Olsen, Michael
        Hallett, Eiliv Lund, Lars Ailo Bongo.  
        \\ 
        The 8th ACM Conference on Bioinformatics, Computational Biology, and
        Health Informatics (ACM BCB) 2017.

    \item 
        \emph{Interactions between the tumor and the blood systemic response of
        breast cancer patients.}
        \\ 
        Vanessa Dumeaux, \textbf{Bjørn Fjukstad}, Hans E Fjosne, Jan-Ole
        Frantzen, Marit Muri Holmen, Enno Rodegerdts, Ellen Schlichting,
        Anne-Lise Børresen-Dale, Lars Ailo Bongo, Eiliv Lund, Michael Hallett.
        \\ 
        PLoS Computational Biology 2017.

    \item \emph{A Review of Scalable Bioinformatics Pipelines} 
        \\
        \textbf{Bjørn Fjukstad}, Lars Ailo Bongo.
        \\ 
        Data Science and Engineering 2017.

        

    \item \emph{nsroot: Minimalist Process Isolation Tool Implemented With Linux
        Namespaces.}
        \\
        Inge Alexander Raknes, \textbf{Bjørn Fjukstad}, Lars Ailo Bongo 
        \\
        NIK 2017. 


    \item \emph{Transcription factor PAX6 as a novel prognostic factor and
        putative tumour suppressor in non-small cell lung cancer} 
        \\
        Yury Kiselev, Sigve Andersen, Charles Johannessen, \textbf{Bjørn
        Fjukstad}, Karina Standahl Olsen, Helge Stenvold, Samer Al-Saad, Tom
        Dønnem, Elin Richardsen, Roy M Bremnes, and Lill-Tove Rasmussen Busund. 
        \\
        Scientific Reports 2018. 

    \item \emph{Reproducible Data Analysis Pipelines in Precision Medicine}
        \\
        \textbf{Bjørn Fjukstad}, Vanessa Dumeaux, Michael Hallett, Lars Ailo
        Bongo
        \\
        Conference TBA 2018. 
        
\end{itemize} 




\section{Dissertation Plan} 



% et kapitelle per paper (kopi tekst) 
% ekstra related work: 
% diskusjon, coclu
