% Thesis project two parts: 
% 1: Systems for exploring genomic data. Advanced statistical analyses from the
% web-browser or any other platform. 
% 2:  Analysis pipelines for large genomic datasets. Provenance of data. 
% 
% Common: 
% - Containers
% - Reproducibility
%     - 1: An application is an R package with the analyses. Do not need to
%     pre-compute results, can call functions etc. directly from the app. Build an
%     app as containers that can be shared. 
%     - 2: An analysis is a list of tools with 
%     input/output data. We package each tool in a container (snapshot-ish) 

% observation
There is a rapid growth in the number of available biological datasets from the
decreaseing cost of data collection. This brings the opportunity to gain novel
insights in the underlying biological mechanisms in the development and
progression of diseases such as cancer. The growing number of datasets has led
to the availablility of a wealth of software packages and systems to explore and
analyze these datasets. However, these systems typically target only small
portions of the deep analysis pipeline required to transform the raw data into
interpretable results. While the tools are used to provide novel insights in
diseases, there is little emphasis on reporting and sharing information about
tool versions, input parameters, and other information that can help others use
the same methods on their own datasets. This leads to unneccesary difficilties
to reuse known methods, and difficulties in reproducing analyses.  This has led
to unrealized potential in the growing number of available datasets. 

% LA
% - Disse er skrevet fra “utvikler ståsted”, og med en “task” form. Hva er
%   challenges for brukerne? Og hvem er det?
% - Det er for vanskelig å gjøre data analyse, så det tar for lang tid å få ut
%   resultater.
% - Lang vei fra instrument til biologisk meningsfulle resultater.
% - Stor variasjon i studie design, data typer, og biologiske spørsmål

% “An overview” til hvem?
% for vanskelig å analysere
% ta det opp ett nivå (hvor ønsker å gjøre det)
% hvorfor 
% hva er problemet
% 1) for mye tid pipelines (før)
% 2) etter analyser ikke huske 
% 3) lang vei og mennesker fra instr til tolkning
% 4) stor variasjon i analyser (vanskelig å 


% challenge
\begin{enumerate*}[label=(\roman*)]
    \item I
\end{enumerate*} 



% \begin{enumerate*}[label=(\roman*)]
%     \item Providing an overview of the full data transformations from raw data
%         to interpretable results; 
%     \item Sharing analysis pipelines and exploration tools across software
%         platforms and research groups; 
%     \item Providing full provenance of data, tools, and tool parameters; 
%     \item Develop end-user applications that interface with advanced
%         statistical analyses required to explore biological datasets.  
% \end{enumerate*} 

\emph{Related work}:
% 
There are a wealth of related work and systems. To build analysis pipelines we
have systems such as
\begin{enumerate*}[label=(\roman*)]
    \item CWL and its implementations: cwl\_runner, Arvados, Rabix, Toil, Galaxy,
        AWE; and 
    \item Pahchyderm. 
\end{enumerate*}
When the data is analyzed and ready for further exploration, we have systems
such as
\begin{enumerate*}[label=(\roman*)]
    \item OpenCPU;
    \item RStudio and Shiny; 
    \item Renjin an R interpreter built on top of the JVM. 
\end{enumerate*}
 

\emph{Our solution}: 
% unified/hollistic: unified approach/model 
% kvik/walr (data straming/ of open source systems including Apache Spark as the
% common foundation; Shark for SQL processing; and Spark Streaming for
% distributed stream processing) 
% nowac-pakken 

To provide an overview of the statistical analyses from raw data to
interpretable results we propose our tool \emph{walrus}. It lets users create
and run analysis pipelines in bioinformatics to e.g. analyze high-throughput
sequencing datasets. In addition, it tracks full provenance of the input,
intermediate, and output data, as well as tool parameters. With \emph{walrus} we
have successfully built analysis pipelines to detect somatic mutations in breast
cancer patients. 

To develop applications that interface with the underlying statistical analyses
we have built \emph{Kvik}. Kvik allows applications written in any modern
programming language to interface with the wealth of bioinformatics packages in
the R programming language, as well as information available through online
databases. We have used Kvik to develop the MIxT system for exploring and
comparing transcriptional profiles from blood and tumor samples in breast cancer
patients. 

\emph{Advantages of our solution}:
% bruk challenges for å si hvorfor vår er bedre

Our analysis solution has the benefit that it integrates with modern version
control systems to provide provenance information on datasets. It also runs on a
wide range of software platforms and is targeted to the compute infrastructure
found in hospital environment. 

Our solution to build data explorations provides the benefit of developers being
able to develop applications in any programming language. It also facilitates
the reuse and interface with the wealth of statistical packages directly from a
user-interface rather than analysis scripts. 

\emph{Methods}:
% We implement the RDD architecture in a stack of open source systems including
% Apache Spark as the common foundation; evaluation: 
% Litt usikker her LA? 

\emph{Long term contributions}:
Discuss why our approach works out nicely? Benefit for future use? Clinical
apps? 

\emph{Thesis statement}:
% Denne må jeg ha hjelp med! :) 
% hva hollistic can ... support diverse distr. comp.
A holistic approach to data analysis and exploration of high-throughput
dataasets in bioinformatics??? 

% alle stegene: pre-proessering, analyser, vise apps 
% hovedresultat: kapt 2
% exp: bcb-paperet 
% pipppelin epaper

\section{List of papers} 
\begin{itemize}
    \item
        \emph{Kvik: three-tier data exploration tools for flexible analysis of
        genomic data in epidemiological studies}
        \\
        \textbf{Bjørn Fjukstad}, Karina Standahl Olsen, Mie Jareid, Eiliv Lund,
        Lars Ailo Bongo. 
        \\ 
        F1000Research 2015.
        
    \item 
        \emph{Building Applications For Interactive Data Exploration In Systems
        Biology.}
        \\
        \textbf{Bjørn Fjukstad}, Vanessa Dumeaux, Karina Standahl Olsen, Michael
        Hallett, Eiliv Lund, Lars Ailo Bongo.  
        \\ 
        The 8th ACM Conference on Bioinformatics, Computational Biology, and
        Health Informatics (ACM BCB) 2017.

    \item 
        \emph{Interactions between the tumor and the blood systemic response of
        breast cancer patients.}
        \\ 
        Vanessa Dumeaux, \textbf{Bjørn Fjukstad}, Hans E Fjosne, Jan-Ole
        Frantzen, Marit Muri Holmen, Enno Rodegerdts, Ellen Schlichting,
        Anne-Lise Børresen-Dale, Lars Ailo Bongo, Eiliv Lund, Michael Hallett.
        \\ 
        PLoS Computational Biology 2017.

    \item \emph{A Review of Scalable Bioinformatics Pipelines} 
        \\
        \textbf{Bjørn Fjukstad}, Lars Ailo Bongo.
        \\ 
        Data Science and Engineering 2017.

    \item \emph{Reproducible Data Analysis Pipelines in Precision Medicine}
        \\
        \textbf{Bjørn Fjukstad}, Vanessa Dumeaux, Michael Hallett, Lars Ailo
        Bongo
        \\
        Conference TBA 2018. 


    \item \emph{nsroot: Minimalist Process Isolation Tool Implemented With Linux
        Namespaces.}
        \\
        Inge Alexander Raknes, \textbf{Bjørn Fjukstad}, Lars Ailo Bongo 
        \\
        NIK 2017. 
        
\end{itemize} 

\section{Problems with Data Analysis and Exploration in Bioinformatics} 
    List of things that we want to fix. 
    
\section{The Container-based Data Analysis Model (CDAM)} 
    The Solution (?) 

\section{Use of the CDAM} 

\textbf{walrus}
\textbf{kvik} 
\textbf{mixt} 

\section{Summary of Results} 

\section{Dissertation Plan} 


% et kapitelle per paper (kopi tekst) 
% ekstra related work: 
% diskusjon, coclu
