% observation
There is a rapid growth in the number of available biological datasets due to
the decreaseing cost of data collection.  This brings opportunities to gain
novel insights to the underlying biological mechanisms in the development and
progression of diseases such as cancer.  The wide range of different biological
datasets has led to the development of a wealth of software packages and systems
to explore and analyze these datasets.  However, these tools typically target
only small portions of the deep analysis pipeline required to transform the raw
data into interpretable results. While the tools are used to provide novel
insights in diseases, there is little emphasis on reporting and sharing
information about tool versions, input parameters, and other information that
can help others use the same methods on their own datasets. This leads to
unneccesary difficilties to reuse known methods, and difficulties in reproducing
analyses, which leads to unrealized potential for scientific insights and
commercial use in the available datasets. 

% challenge
There are several challenges for researchers to analyze and explore biological
datasets. These challenges are common for large datasets such as high-throughput
sequencing data that require long-running, deep analysis pipelines, as well as
smaller datasets, such as microarray data, that require complex, but
short-running analysis pipelines.  The first is the time and knowledge required
to find and set up the necessary analysis tools to start analyzing a modern
biological dataset.  The second is ensuring the correct input parameters, tool
versions, database versions, and dataset versions when analyzing, and reporting
analysis results to enable reproducible science. The final challenge is the need
for flexible and highly configurable analysis tools to support the large
variation in study design and data types.

% related 
As a result, there are a wealth of approaches and systems to enable analysis of
the complex biological data. To develop deep analysis pipelines,
Galaxy\cite{galaxy} has for a long time provided a simple interface to set up
and execute analysis pipelines for genomic datasets. However, the Galaxy system
is less effective for explorative and flexible
analyses.\cite{spjuth2015experiences} Pachyderm is a system for
developing more flexible analyses that support comparison of pipeline runs from
different workflow configurations and datasets.\cite{pachyderm} However it has
yet to see wide-spread adpotion in Bioinformatics.  New initiatives such as the
\gls{cwl} provide users a standardized way of describing an analysis pipeline
and has multiple implementations such as the reference implementation
cwl\_runner,\footnote{\url{github.com/common-workflow-language/cwltool}}
Arvados,\cite{arvados} Rabix,\cite{rabix} Toil,\cite{toil} Galaxy,\cite{galaxy}
and AWE.\cite{awe} While these systems provide system for batch-processing of
large datasets, systems such as Shiny and OpenCPU provide interactive interfaces
to the R programming language and the many packages for biological data
processing in Bioconductor.  With the addition of new datasets and methods every
year, it seems that analysis of biological data requires a wide array of
different tools and systems.

% our solution
This dissertation argues that, instead, we can design a unified approach that
integrates disperate tools and data into fully reproducible biological data
analysis frameworks.  In particular, we show how software container
technologies, such as Docker, provide the necessary foundation to build
reproducible environments for any analysis pipeline, as well as a suitable
environent to package an entire data expliration application. 

The resulting approach as several key advantages over current systems: 
\begin{itemize} 
    \item It simplifies sharing of 
    \item focus on analyses not solving dependencies and installing
\end{itemize} 



We implement walrus/kvik... 

% We implement the RDD architecture in a stack of open source systems including
% Apache Spark as the common foundation; evaluation: 
From a longer-term perspective, ....

\emph{Thesis statement}:
A common development model based on software container infrastructure can
efficiently provide reproducible and easy to use environments to develop
applications for exploring biological datasets. 

\section{Problems with Data Analysis and Exploration in Bioinformatics} 
    List of things that we want to fix. 

%  no way I'm keeping the acronym    
\section{The Container Centric Analysis Model (CDAM)} 
    The Solution (?) 

\section{Use of the CDAM} 

To provide an overview of the statistical analyses from raw data to
interpretable results we propose our tool \emph{walrus}. It lets users create
and run analysis pipelines in bioinformatics to e.g. analyze high-throughput
sequencing datasets. In addition, it tracks full provenance of the input,
intermediate, and output data, as well as tool parameters. With \emph{walrus} we
have successfully built analysis pipelines to detect somatic mutations in breast
cancer patients. 

To develop applications that interface with the underlying statistical analyses
we have built \emph{Kvik}. Kvik allows applications written in any modern
programming language to interface with the wealth of bioinformatics packages in
the R programming language, as well as information available through online
databases. We have used Kvik to develop the MIxT system for exploring and
comparing transcriptional profiles from blood and tumor samples in breast cancer
patients. 


\textbf{walrus}
\textbf{kvik} 
\textbf{mixt} 

\section{Summary of Results} 

\section{List of papers} 
This section contains a list of papers along with short descriptions and my
personal contribution to each paper. 
\capstartfalse
\begin{table}[H]
    \centering
    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & Kvik: three-tier data exploration tools for flexible analysis
         of genomic data in epidemiological studies \\ \hline
         
         Authors & \textbf{Bjørn Fjukstad}, Karina Standahl Olsen, Mie Jareid,
         Eiliv Lund, and Lars Ailo Bongo \\ \hline
         
         Description & The initial description of Kvik, and how we used it to
         implement Kvik Pathways, a web application for browsing biologicap
         pathway maps integrated with gene expression data from the \gls{nowac}
         cohort. 
         \\ \hline
         
         Contribution & Designed, implemented, and deployed Kvik and Kvik
         Pathways. Evaluated the system and wrote the manuscript. \\ \hline
         
         Publication date & 15 March 2015 \\ \hline 

         Publication venue & F1000 \\ \hline
         
         Citation & \cite{fjukstad2015kvik} \bibentry{fjukstad2015kvik} \\
         \hline 
    \end{tabular}
    \label{p1}

    \hfill 

    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & Building Applications For Interactive Data Exploration In
         Systems Biology. \\ \hline
         
         Authors & \textbf{Bjørn Fjukstad}, Vanessa Dumeaux, Karina
         Standahl Olsen, Michael Hallett, Eiliv Lund, and Lars Ailo Bongo.  \\
         \hline
         
         Description & Describes how we further developed the ideas from Paper 1
         into an approach that we used to build the \gls{mixt} web application. 
         \\ \hline
         
         Contribution & 
         Designed, implemented, and deployed Kvik and the \gls{mixt} web
         application.  Evaluated the system and wrote the manuscript. 
         \\ \hline
         
         Publication date & 20 August 2017. \\ \hline  

         Publication venue & The 8th ACM Conference on Bioinformatics,
         Computational Biology, and Health Informatics (ACM BCB) August 20–23, 2017.  \\
         \hline
         
         Citation & \cite{fjukstad2017building} \bibentry{fjukstad2017building}
         \\ \hline 
    \end{tabular}
    \label{p2}
    
\end{table}

\begin{table}[H]
    \centering
    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & Interactions Between the Tumor and the Blood Systemic Response
         of Breast Cancer Patients \\ \hline
         
         Authors & Vanessa Dumeaux, \textbf{Bjørn Fjukstad}, Hans E Fjosne,
         Jan-Ole Frantzen, Marit Muri Holmen, Enno Rodegerdts, Ellen
         Schlichting, Anne-Lise Børresen-Dale, Lars Ailo Bongo, Eiliv Lund,
         Michael Hallett.  \\ \hline
         
         Description & Describes the \gls{mixt} system which enables
         identification of genes and pathways in the primary tumor that are tightly
         linked to genes and pathways in the patient \gls{sr}. 
         \\ \hline
         
         Contribution & 
         Designed, implemented, and deployed the \gls{mixt} web application.
         Helped write the manuscript. 
         \\ \hline
         
         Publication date & 28 September 2017. \\ \hline  

         Publication venue &  PLoS Computational Biology \\ \hline
         
         Citation & \cite{dumeaux2017interactions}
         \bibentry{dumeaux2017interactions}
         \\ \hline 
    \end{tabular}
    \label{p3}
    
    \hfill 

    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & A Review of Scalable Bioinformatics Pipelines \\ \hline
         
         Authors & \textbf{Bjørn Fjukstad}, Lars Ailo Bongo. \\ \hline
         
         Description & This review survey several scalable bioinformatics
         pipelines and compare their design and their use of underlying
         frameworks and infrastructures.      \\ \hline
         
         Contribution & 
         Wrote the manuscript.  \\ \hline
         
         Publication date & 23 October 2017 \\ \hline  

         Publication venue & Data Science and Engineering 2017. \\ \hline
         
         Citation & \cite{fjukstad2017review} \bibentry{fjukstad2017review} \\
         \hline 
    \end{tabular}
    \label{p4}
\end{table}
\begin{table}[H]
    \centering
    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & nsroot: Minimalist Process Isolation Tool Implemented With
         Linux Namespaces.  \\ \hline
         
         Authors & Inge Alexander Raknes, \textbf{Bjørn Fjukstad}, Lars Ailo Bongo. \\ \hline
         
         Description & Describes a tool for process isolation built using Linux
         namespaces.          \\ \hline
         
         Contribution & Helped write the
         manuscript, specifically to the literature review and related works.
         \\ \hline
         
         Publication date & 26 November 2017 \\ \hline  

         Publication venue & Norsk Informatikkonferanse 2017. \\ \hline
         
         Citation & \cite{fjukstad2017review} \bibentry{fjukstad2017review} \\
         \hline 
    \end{tabular}
    \label{p5}
\end{table}
\begin{table}[H]
    \centering
    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & Transcription factor PAX6 as a novel prognostic factor and
        putative tumour suppressor in non-small cell lung cancer \\ \hline
         
         Authors & Yury Kiselev, Sigve Andersen, Charles Johannessen, 
         \textbf{Bjørn Fjukstad}, Karina Standahl Olsen, Helge Stenvold, Samer
         Al-Saad, Tom Dønnem, Elin Richardsen, Roy M Bremnes, and Lill-Tove
         Rasmussen Busund.\\ \hline
         
         Description & 
         \\ \hline
         
         Contribution & 
         \\ \hline
         
         Publication date & 22 March 2018 \\ \hline  

         Publication venue & Scientific Reports 2018. \\ \hline
         
         Citation & \cite{kiselev2018transcription} \bibentry{kiselev2018transcription} \\
         \hline 
    \end{tabular}
    \label{p6}
\end{table}
\begin{table}[H]
    \centering
    \begin{tabular}{ | l | p{9.5cm} | }
    \hline
         Title & Reproducible Data Analysis Pipelines in Precision Medicine \\
         \hline
         
         Authors &  \textbf{Bjørn Fjukstad}, Vanessa Dumeaux, Michael Hallett,
         Lars Ailo Bongo\\ \hline
         
         Description & This paper outlines how we used the container centric
         development model to build walrus. 
         \\ \hline
         
         Contribution & Design, implementation and evaluation of walrus. Wrote
         the manuscript. 
         \\ \hline
         
         Publication date & TBA \\ \hline  

         Publication venue & TBA \\ \hline
         
         Citation & \cite{walrus} \bibentry{walrus} \\
         \hline 
    \end{tabular}
    \label{p6}
\end{table}

% kapittel 2
% pippelinen + review + nowac
% generelt + kvinner og kreft
% kapittel 3: walrus 
% kapittel 4: kvik/mixt 


% Husk historien: Fra rå data gjennom komplekse analyse pipeliner og helt til
% forskere kan svømme rundt i resultatene. 

\section{Dissertation Plan} 
This thesis is organized as follows. Chapter 2 describes the characteristics of
state-of-the-art biological datasets, the analysis required to extract knowledge
from these, and the available tools and analysis frameworks. Chapter 3 describes
in detail how we use a container centric development model to build a tool,
walrus, to develop and execute deep analysis pipelines. In Chapter 4 we describe
how we used the same model to develop applications to interactively explore
results from statistical analyses.  Finally, Chapter 5 concludes the work and
discusses future directions. 

